<a href="https://colab.research.google.com/github/odevhesabi1/Odevler/blob/main/%C3%96dev%20-%20Modelin%20E%C4%9Fitim%20Performans%C4%B1n%C4%B1%20%C3%96l%C3%A7me.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [3]:
import pandas as pd
import numpy as np
import seaborn as sns
import statsmodels.api as sm
from sklearn import linear_model
import matplotlib.pyplot as plt

weather_df = pd.read_csv("https://djl-lms-assets.s3.eu-central-1.amazonaws.com/datasets/weatherHistory.csv")
weather_df.head()

/usr/local/lib/python3.7/dist-packages/statsmodels/tools/_testing.py:19: FutureWarning: pandas.util.testing is deprecated. Use the functions in the public API at pandas.testing instead.
  import pandas.util.testing as tm


Formatted Date  ...                      Daily Summary
0  2006-04-01 00:00:00.000 +0200  ...  Partly cloudy throughout the day.
1  2006-04-01 01:00:00.000 +0200  ...  Partly cloudy throughout the day.
2  2006-04-01 02:00:00.000 +0200  ...  Partly cloudy throughout the day.
3  2006-04-01 03:00:00.000 +0200  ...  Partly cloudy throughout the day.
4  2006-04-01 04:00:00.000 +0200  ...  Partly cloudy throughout the day.

[5 rows x 12 columns]

In [4]:
#Önceki derste olduğu gibi, hedef değişkeninizin görünür sıcaklık ve sıcaklık arasındaki fark olduğu bir doğrusal regresyon modeli oluşturun. Açıklayıcı değişkenler olarak nem ve rüzgar hızı kullanın. 
#Şimdi, modelinizi OLS kullanarak tahmin edin. R-kare ve ayarlanmış R-kare değerleri nelerdir? Tatminkar olduklarını düşünüyor musunuz? Açıklayın.

weather_df["TemperatureDifference"] = weather_df["Apparent Temperature (C)"] - weather_df["Temperature (C)"]
weather_df.head()

#Hedef değişken negatif değerlere sahip, bu modeli etkiler mi ?
X = weather_df[["Humidity", "Wind Speed (km/h)"]]
Y = weather_df["TemperatureDifference"]

X = sm.add_constant(X)
results = sm.OLS(Y,X).fit()
results.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                              OLS Regression Results                             
=================================================================================
Dep. Variable:     TemperatureDifference   R-squared:                       0.288
Model:                               OLS   Adj. R-squared:                  0.288
Method:                    Least Squares   F-statistic:                 1.949e+04
Date:                   Sat, 22 May 2021   Prob (F-statistic):               0.00
Time:                           20:16:08   Log-Likelihood:            -1.7046e+05
No. Observations:                  96453   AIC:                         3.409e+05
Df Residuals:                      96450   BIC:                         3.409e+05
Df Model:                              2                                         
Covariance Type:               nonrobust                                         
=====================================================================================
                        coef    std err          t      P>|t|      [0.025      0.975]
-------------------------------------------------------------------------------------
const                 2.4381      0.021    115.948      0.000       2.397       2.479
Humidity             -3.0292      0.024   -126.479      0.000      -3.076      -2.982
Wind Speed (km/h)    -0.1193      0.001   -176.164      0.000      -0.121      -0.118
==============================================================================
Omnibus:                     3935.747   Durbin-Watson:                   0.264
Prob(Omnibus):                  0.000   Jarque-Bera (JB):             4613.311
Skew:                          -0.478   Prob(JB):                         0.00
Kurtosis:                       3.484   Cond. No.                         88.1
==============================================================================

Warnings:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

R değerlerimiz 0.288, bu da modelin toplam varyansın çoğunu açıklayamadığını gösterir. ( Bu arada p değerleri de istatistiksel olarak anlamlı ama p değerlerinin tek başına modelin iyi bir model olduğunu kanıtlamadığını R değerlerimizin düşük olmasından anlayabiliriz. )





In [5]:
#İlk modele ek açıklayıcı değişken olarak görünürlük ekleyin ve tahmin edin. R-kare arttı mı? Ayarlanmış R-kare değeri ne oldu? Tabloda ortaya çıkan farklılıkları, 
#ayarlanmış R-kare içindeki iyileşme açısından etkileşim terimi ve görünürlük ile karşılaştırın. Hangisi daha kullanışlı?

#ETKİLEŞİMLİ ÖZELLİK İÇEREN MODEL 

weather_df["Hum_and_Wind"] = weather_df["Humidity"] * weather_df["Wind Speed (km/h)"]

X = weather_df[["Humidity", "Wind Speed (km/h)", "Hum_and_Wind"]]
Y = weather_df["TemperatureDifference"]

X = sm.add_constant(X)
results = sm.OLS(Y,X).fit()
results.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                              OLS Regression Results                             
=================================================================================
Dep. Variable:     TemperatureDifference   R-squared:                       0.341
Model:                               OLS   Adj. R-squared:                  0.341
Method:                    Least Squares   F-statistic:                 1.666e+04
Date:                   Sat, 22 May 2021   Prob (F-statistic):               0.00
Time:                           20:44:47   Log-Likelihood:            -1.6669e+05
No. Observations:                  96453   AIC:                         3.334e+05
Df Residuals:                      96449   BIC:                         3.334e+05
Df Model:                              3                                         
Covariance Type:               nonrobust                                         
=====================================================================================
                        coef    std err          t      P>|t|      [0.025      0.975]
-------------------------------------------------------------------------------------
const                 0.0839      0.033      2.511      0.012       0.018       0.149
Humidity              0.1775      0.043      4.133      0.000       0.093       0.262
Wind Speed (km/h)     0.0905      0.002     36.797      0.000       0.086       0.095
Hum_and_Wind         -0.2971      0.003    -88.470      0.000      -0.304      -0.291
==============================================================================
Omnibus:                     4849.937   Durbin-Watson:                   0.262
Prob(Omnibus):                  0.000   Jarque-Bera (JB):             9295.404
Skew:                          -0.378   Prob(JB):                         0.00
Kurtosis:                       4.320   Cond. No.                         193.
==============================================================================

Warnings:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

In [6]:
#VISIBILITY İÇEREN MODEL

X = weather_df[["Humidity", "Wind Speed (km/h)", "Visibility (km)"]]
Y = weather_df["TemperatureDifference"]

X = sm.add_constant(X)
results = sm.OLS(Y,X).fit()
results.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                              OLS Regression Results                             
=================================================================================
Dep. Variable:     TemperatureDifference   R-squared:                       0.304
Model:                               OLS   Adj. R-squared:                  0.303
Method:                    Least Squares   F-statistic:                 1.401e+04
Date:                   Sat, 22 May 2021   Prob (F-statistic):               0.00
Time:                           20:45:24   Log-Likelihood:            -1.6938e+05
No. Observations:                  96453   AIC:                         3.388e+05
Df Residuals:                      96449   BIC:                         3.388e+05
Df Model:                              3                                         
Covariance Type:               nonrobust                                         
=====================================================================================
                        coef    std err          t      P>|t|      [0.025      0.975]
-------------------------------------------------------------------------------------
const                 1.5756      0.028     56.605      0.000       1.521       1.630
Humidity             -2.6066      0.025   -102.784      0.000      -2.656      -2.557
Wind Speed (km/h)    -0.1199      0.001   -179.014      0.000      -0.121      -0.119
Visibility (km)       0.0540      0.001     46.614      0.000       0.052       0.056
==============================================================================
Omnibus:                     3833.895   Durbin-Watson:                   0.279
Prob(Omnibus):                  0.000   Jarque-Bera (JB):             4584.022
Skew:                          -0.459   Prob(JB):                         0.00
Kurtosis:                       3.545   Cond. No.                         131.
==============================================================================

Warnings:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

Etkileşimli terim içeren ilk model, visibility içeren modele göre daha açıklayıcı. Ayrıca yine ilk modelin F istatistiği daha yüksek. 

Bu nedenle ilk model daha kullanılışlı.

In [13]:
# AIC ve BIC puanlarına göre yukarıdaki üç modelden en iyisini seçin. Mentor ile gerekçenizi tartışarak seçiminizi doğrulayın.

#1. MODEL İÇİN AIC BIC DEĞERLERİ
print("1.model için değerler: \n AIC:	3.409e+05 \n BIC:	3.409e+05 \n  ")

#2. MODEL İÇİN AIC BIC DEĞERLERİ
print("2.model için değerler: \n AIC:	3.334e+05 \n BIC:	3.334e+05  \n ")

#3. MODEL İÇİN AIC BIC DEĞERLERİ
print("3.model için değerler: \n AIC:	3.388e+05 \n BIC:	3.388e+05  \n ")


1.model için değerler: 
 AIC:	3.409e+05 
 BIC:	3.409e+05 
  
2.model için değerler: 
 AIC:	3.334e+05 
 BIC:	3.334e+05  
 
3.model için değerler: 
 AIC:	3.388e+05 
 BIC:	3.388e+05  
 


2. model için AIC ve BIC değerleri daha düşük. Bu sonuç, 2. modelin R değerlerinin daha yüksek olması ile de uyumlu.

3 model arasından, 2.model daha kullanışlı.

EV FİYATLARI MODELİNİ, VERİSETİNİ UYGUN HALE GETİREMEDİĞİM İÇİN YAPAMADIM.